python/mxnet/model.py

```diff
@@ -113,25 +127,36 @@ def _update_params_on_kvstore(param_arrays, grad_arrays, kvstore, param_names):
         # push gradient, priority is negative index
         kvstore.push(name, grad_list, priority=-index)
         # pull back the weights
-        kvstore.pull(name, arg_list, priority=-index)
+        if _contains_non_default_storage(arg_list):
```

Comment:
If the arg_list does contain what you described, yes, pull is skipped as you said. However, I failed to see why the arg_list for **one variable** will contain a mixture of sparse and non-sparse values if you do data parallel with executor group. I don't know how to interpret why such case (mixture of storage type) will happen. Do you have any use case in mind?
I probably should explicit assert this assumption in the code.